source/common/http/conn_manager_impl.cc

```diff
@@ -1621,15 +1621,13 @@ void ConnectionManagerImpl::FilterManager::sendLocalReplyViaFilterChain(
   Utility::sendLocalReply(
       state_.destroyed_,
       Utility::EncodeFunctions{
+          modify_headers,
           [this](ResponseHeaderMap& response_headers, Code& code, std::string& body,
                  absl::string_view& content_type) -> void {
             active_stream_.connection_manager_.config_.localReply().rewrite(
                 request_headers_.get(), response_headers, stream_info_, code, body, content_type);
           },
           [this, modify_headers](ResponseHeaderMapPtr&& headers, bool end_stream) -> void {
-            if (modify_headers != nullptr) {
```

Comment:
hmmm, this is clearly an improvement but also clearly causes L7 changes (one fix that we know of, but maybe someone depends on the order?)  I'd lean towards runtime guarding.

Either way we definitely want a release note about the bugfix.

Reply:
I think the change is isolated within this function: ConnectionManagerImpl::ActiveStream::sendLocalReply

Here are the steps in the Utility::sendLocalReply()
1) new response_headers
2) local_reply.write(response_headers)
3) response_headers.set_content_type() + content_length() for the local reply_body
4) modify_headers(response_headers) // which may override content_type and content_length.

This change just moved 4) to be before 2).   I think this change should be local and safe.   Could you elaborate the L7 changes you are worrying?

Reply:
I saw the problem from an integration test failure. 

IpVersions/IntegrationTest.RouterDirectResponse/IPv4
test/integration/integration_test.cc:180: Failure
Expected equality of these values:
  "text/html"
    Which is: 0x3af299
  response->headers().getContentTypeValue()
    Which is: "text/plain"

RouterDirectResponse is also calling sendLocalReply.   It may have response_headers to add.  
With this change,  If Direct_Response.respone_headers has "content-length" and "type", they are overridden by Utility::sendLocalReply()

Reply:
There is a conflict between "modify_headers" in this [API](https://github.com/envoyproxy/envoy/blob/master/include/envoy/http/filter.h#L340) and "local_reply_config" in [HCM](https://github.com/envoyproxy/envoy/blob/e8216a8cf79c54e3e0a77ab729ebf27f4e79eb1b/api/envoy/extensions/filters/network/http_connection_manager/v3/http_connection_manager.proto#L535)

When  "local_reply_config" is used,  the reply body and its type is generated by Utility::sendLocalReply(), modify_headers should not override it, otherwise it will be wrong.   

But I also see the usage from Route_Direct_Response where it wants to set content-type for its response body.

Reply:
I think this policy should work:

1) content-length should always be set by Utility::sendLocalReply() since it generates reply body,  it should override one set by `modify_headers`  function.

2) if reply body is modified by `local_reply_config` or `modify_headers` did not set content-type, set its content-type in Utility::sendLocalReply()

2) is to fix IpVersions/IntegrationTest.RouterDirectResponse/IPv4 test failure where redirect_response wants to set content-type as "text/html", but Utility::sendLocalReply() set it as "plain/text"